packages/hierarchy-builder/src/hierarchy-builder/internal/operators/LabelGrouping.ts

```diff
@@ -0,0 +1,112 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { from, mergeMap, Observable, tap, toArray } from "rxjs";
+import { HierarchyNode } from "../../HierarchyNode";
+import { getLogger } from "../../Logging";
+import { createOperatorLoggingNamespace } from "../Common";
+
+const OPERATOR_NAME = "Grouping.ByLabel";
+/** @internal */
+export const LOGGING_NAMESPACE = createOperatorLoggingNamespace(OPERATOR_NAME);
+
+/** @internal */
+export function createLabelGroupingOperator() {
+  return function (nodes: Observable<HierarchyNode>): Observable<HierarchyNode> {
+    return nodes.pipe(
+      log((n) => `in: ${n.label}`),
+      // need all nodes in one place to group them
+      toArray(),
+      // group all nodes and format into a nodes observable
+      mergeMap((resolvedNodes) => {
+        const obs = from(createLabelGroups(resolvedNodes));
+        return obs;
+      }),
+      log((n) => `out: ${n.label}`),
+    );
+  };
+}
+
+function createLabelGroups(nodes: HierarchyNode[]): HierarchyNode[] {
+  if (nodes.length === 0) {
+    return nodes;
+  }
+  const [firstNode, firstHasChanged] = createLabelGroupsIfClassGroupingNode(nodes[0]);
+  const outputNodes: HierarchyNode[] = [firstNode];
+  let hasChanged = firstHasChanged;
+
+  for (let i = 1; i < nodes.length; ++i) {
+    const [currentNode, currentHasChanged] = createLabelGroupsIfClassGroupingNode(nodes[i]);
+    if (currentHasChanged) {
+      hasChanged = currentHasChanged;
+    }
+
+    if (currentNode.label === outputNodes[outputNodes.length - 1].label) {
+      const lastOutputNode = outputNodes[outputNodes.length - 1];
+      if (HierarchyNode.isLabelGroupingNode(lastOutputNode) && Array.isArray(lastOutputNode.children)) {
+        if (currentNode.params?.groupByLabel) {
+          lastOutputNode.children.push(currentNode);
+        } else {
+          outputNodes.push(currentNode);
+          [outputNodes[outputNodes.length - 1], outputNodes[outputNodes.length - 2]] = [
+            outputNodes[outputNodes.length - 2],
+            outputNodes[outputNodes.length - 1],
+          ];
```

Comment:
maybe just insert straight into the right location?

Reply:
suggestion also applies to a code piece below

Reply:
Doing this instead:
outputNodes.push(outputNodes[outputNodes.length - 1]);
outputNodes[outputNodes.length - 2] = currentNode;

Reply:
why not `outputNodes.splice(outputNodes.length - 1, 0, currentNode)`

Reply:
You are right, splice is simpler. Updated to your suggestion.